chain/src/chain.rs

```diff
@@ -807,34 +807,41 @@ impl Chain {
 		while let Ok(header) = current {
 			// break out of the while loop when we find a header common
 			// between the header chain and the current body chain
-			if header.height <= body_head.height {
-				if let Ok(_) = self.is_on_current_chain(&header) {
-					break;
-				}
+			if let Ok(_) = self.is_on_current_chain(&header) {
+				oldest_height = header.height;
+				oldest_hash = header.hash();
+				break;
 			}
 
-			oldest_height = header.height;
-			oldest_hash = header.hash();
-			if let Some(hs) = hashes {
-				hs.push(oldest_hash);
-			}
 			current = self.get_previous_header(&header);
 		}
 
+		// Go through the header chain reversely to get all those headers after oldest_height.
+		if let Some(hs) = hashes {
+			let mut current = self.get_block_header(&header_head.last_block_h);
```

Comment:
I see we are redefining current in this PR here, but I don't think we need to do this. Earlier this is `body_head` and now this is redefined based on `header_head`. 

I think it should consistently be based on a header from the db starting at `header_head` and working back until we find one that also refers to a full block in our db.